src/EventListener/DataContainer/TagListener.php

```diff
@@ -75,4 +76,108 @@ private function getManager(string $alias): ManagerInterface
     {
         return $this->registry->get($alias);
     }
+
+    /**
+     * Auto-generate the tag alias if it has not been set yet
+     *
+     * @param mixed         $value
+     * @param DataContainer $dc
+     *
+     * @return mixed
+     *
+     * @throws \Exception
+     */
+    public function generateAlias($value, DataContainer $dc)
+    {
+        $autoAlias = false;
+
+        // Generate alias if there is none
+        if ($value == '')
+        {
+            $autoAlias = true;
+            $value     = \StringUtil::generateAlias($dc->activeRecord->name);
+        }
+
+        $alias = \Database::getInstance()->prepare("SELECT id FROM tl_cfg_tag WHERE alias=? AND source=?")
+            ->execute($value, $dc->activeRecord->source);
+
+        // Check whether the alias exists
+        if ($alias->numRows > 1 && !$autoAlias)
+        {
+            throw new \Exception(sprintf($GLOBALS['TL_LANG']['ERR']['aliasExists'], $value));
+        }
+
+        // Add ID to alias
+        if ($alias->numRows && $autoAlias)
+        {
+            $value .= '-' . $dc->id;
+        }
+
+        return $value;
+    }
+
+    /**
+     * Automatically generate the folder URL aliases
+     *
+     * @param array $buttons
+     * @param \DataContainer $dc
+     *
+     * @return array
+     */
+    public function addAliasButton($buttons, \DataContainer $dc)
```

Comment:
This whole method needs to be reworked to use strict types and comparisons, database **service**, etc.